File: LC/Medium/633.py
class Solution:
    def judgeSquareSum(self, c: int) -> bool:
        a = 0
        b = int(sqrt(c))
        while a <= b:
            curr = a * a + b * b
            
            if curr == c: 
                return True
            
            if curr < c:
                a += 1
            else:
                b -= 1

        return False

File: LC/C++/Medium/633.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_set<int> seen;
    int dfs(TreeNode* curr) {
        if (curr == nullptr) {
            return 0;
        }
        int res = dfs(curr->left) + dfs(curr->right) + curr->val;
        seen.insert(res);
        return res;
    }

    bool checkEqualTree(TreeNode* root) {
        int total = dfs(root->left) + dfs(root->right) + root->val;
        return total % 2 == 0 && seen.count(total / 2);
    }
};